src/librustc/ty/context.rs

```diff
@@ -1410,6 +1418,24 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         self.mk_ty(TyAnon(def_id, substs))
     }
 
+    pub fn mk_type_list<I: SliceInternable<Ty<'tcx>,
+                        &'tcx Slice<Ty<'tcx>>>>(self, iter: I) -> I::Output {
+        iter.intern_with(|xs| self.intern_type_list(xs))
+    }
+
+    pub fn mk_substs<I: SliceInternable<Kind<'tcx>,
+                     &'tcx Slice<Kind<'tcx>>>>(self, iter: I) -> I::Output {
+        iter.intern_with(|xs| self.intern_substs(xs))
+    }
+
+    pub fn mk_substs_trait(self,
+                     s: Ty<'tcx>,
+                     t: &[Ty<'tcx>])
```

Comment:
This should probably take `I: SliceInternable`, but then the current code doesn't support passing it to iter::chain, since `SliceInternable` is not an `Iterator`. @eddyb Thoughts on what we should/could do here?